### DISCUSSION TITLE
ValueError('signal only works in main thread')

### DISCUSSION THREAD

**Discussion by maxwass** [Lightning-AI/pytorch-lightning#9589] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Has anyone else run into this error:
`ValueError('signal only works in main thread')`

I'm running a hyper parameter sweep using Weights and Biases's framework.

Running on a GPU on Google Colab which causes all launched runs to fail. Running it locally (Mac OS) prompts 'signal only works in main thread' to be printed to stdout (which also happens on Colab) but it doesn't crash.

Any ideas? It seems people using [Ray with PL](https://github.com/PyTorchLightning/pytorch-lightning/issues/3651) have come across this. The hacky solution presented there (`os.environ['SLURM_JOB_NAME'] = 'bash'`) doesn't work in my case (neither on Mac OS or Colab).

**Comment by borisdayma** [comment]
Hi @maxwass 

Does [this notebook](https://colab.research.google.com/drive/1Tr9dYlwBKk6-LgLKGO8KYZULnguVA992?usp=sharing#scrollTo=ceYIx-8ezA5y) work if you run it locally?

I have [another example](https://github.com/borisdayma/lightning-kitti) but it may be outdated with latest PL versions (haven't tested it recently).

**Reply by maxwass** [reply]
I removed PL and made a dummy hp sweep as such:

```
def train():
   with wandb.init() as run:
      config = wandb.config
      a, b, c = random numbers
      run.log( {"a": a, "b": b, "c": c} )

sweep_config = { "method": "random", "metric": {"name": "a", "goal": "minimize"}, "parameters": {some_parameters} }
sweep_id = wandb.sweep(sweep_config, project='my_project')
wandb.agent(sweep_id=sweep_id, project=project, function=train)
```

On my local machine: MacBook Air (13-inch, Early 2015), macOS Big Sur version 11.2.2, running python 3.7, wandb version 0.12.2. It works, but still prints `signal only works in main thread` on each run.

I also tried it with wandb version 0.12.1 with the same outcome. Will test this on another machine (linux, ubuntu machine), but I believe I will have the same problem.

The main differences I see between this and my desired hp search with my PL model/trainer are: whatever happens in the PL/W&B callback code (in my dummy I directly use the `run` object, and maybe some multi-threading/process stuff with many workers for the dataloaders. Although this is definitely not a comprehensive list.

Any ideas?

**Comment by maxwass** [comment]
Update: My code  successfully runs on an aws server, and my local Mac  printing 'signal only works in main thread', but does not fail. On Colab, it fails with `Run snk3j2ue errored: ValueError('signal only works in main thread')`

**Reply by vanpelt** [reply]
The only place we're using signals is when we're wrapping the terminal console to capture stdout and stderr.  The team is working on some improvements to multiprocessing that should help with this, but in the meantime can you try setting the `WANDB_CONSOLE` environment variable to "off" in your environment and see if the error still occurs?

**Reply by garrett361** [reply]
@vanpelt having this same issue and tried setting WANDB_CONSOLE to off, but it doesn't fix it. Same error. Any other workarounds?

**Reply by garrett361** [reply]
Discovered that it was the upgrade to `pl` v1.5.0 released yesterday which causes the problem: [started thread here](https://github.com/PyTorchLightning/pytorch-lightning/issues/10336#issue-1043728191).  Reverting to v1.4.9 fixes the problem.

**Comment by maxwass** [comment]
Now running on Colab and no longer getting the printed `'signal only works in main thread'` nor the `ValueError('signal only works in main thread')` : I had to port my sweep over to a yaml file and run it with the wandb login/sweep/agent calls from the CLI, not the python API.

**Reply by borisdayma** [reply]
Just letting you know that we are looking into the issue here.

**Reply by garrett361** [reply]
@borisdayma any update on this? Same ValueError issue started happening for me running sweeps in Colab today. Seemed to be working fine one second, then can't sweep at all now. Syncing a single run works fine, though.

**Reply by garrett361** [reply]
Discovered that it was the upgrade to `pl` v1.5.0 released yesterday which causes the problem: [started thread here](https://github.com/PyTorchLightning/pytorch-lightning/issues/10336#issue-1043728191).  Reverting to v1.4.9 fixes the problem.

**Reply by rkechols** [reply]
Same issue of ray tune + pl, and going back to `pytorch-lightning==1.4.9` fixed it for me too 👍

**Reply by max0x7ba** [reply]
> Just letting you know that we are looking into the issue here.

I encounter this error as well with ray tune 1.8.0 and pl.

**Comment by kaushikb11** [comment, marked as answer]
@max0x7ba @borisdayma The issue has been with #10610

**Comment by denisoliveirac** [comment]
A regression was introduced in version 1.5.5 and the problem returned.
This [PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/10610/files#diff-455634aa5542cb67cb1ec373b47aa36cea4a924164006add72afec20a14818c7) solved the problem and was released in version 1.5.3. It also works in version 1.5.4, but in version 1.5.5 the problem returned.

**Reply by garrett361** [reply]
I'm also getting a `ValueError('signal only works in main thread')` message once again, but the runs actually sync up with `wandb` as expected despite the message (setup described in [#10336](https://github.com/PyTorchLightning/pytorch-lightning/issues/10336)).

**Reply by garrett361** [reply]
Hopefully fixed in #11124
